### PR TITLE
fix dp reduction test

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?

Fixes https://github.com/PyTorchLightning/pytorch-lightning/pull/6324#discussion_r586926124

Adds a real test for #6324

- Now runs on 2 GPUs and performs actual reduction (before batch size was 1)
- Test also includes the type conversion check

## Before submitting
- [ ] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [ ] Did you make sure to update the documentation with your changes? (if necessary)
- [ ] Did you write any new necessary tests? (not for typos and docs)
- [ ] Did you verify new and existing tests pass locally with your changes?
- [ ] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [ ] Is this pull request ready for review? (if not, please submit in draft mode)
 - [ ] Check that all items from **Before submitting** are resolved
 - [ ] Make sure the title is self-explanatory and the description concisely explains the PR
 - [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: tests/accelerators/test_dp.py [hunk 1]
```diff
@@ -127,17 +128,56 @@ def test_dp_test(tmpdir):
...
+
+        def add_outputs(self, output, device):
+            output.update({
+                "reduce_int": torch.tensor(device.index, dtype=torch.int, device=device),
+                "reduce_float": torch.tensor(device.index, dtype=torch.float, device=device),
+            })
+
+        def training_step(self, batch, batch_idx):
+            output = super().training_step(batch, batch_idx)
+            self.add_outputs(output, batch.device)
```
Comment (awaelchli) [inline tests/accelerators/test_dp.py:152]: btw, using `batch.device` here because `self.device` seems to be broken for DP

File: tests/accelerators/test_dp.py [hunk 2]
```diff
@@ -127,17 +128,56 @@ def test_dp_test(tmpdir):
...
+    class TestModel(BoringModel):
+
+        def train_dataloader(self):
+            return DataLoader(RandomDataset(32, 64), batch_size=2)
+
+        def val_dataloader(self):
+            return DataLoader(RandomDataset(32, 64), batch_size=2)
+
+        def test_dataloader(self):
+            return DataLoader(RandomDataset(32, 64), batch_size=2)
```
Comment (Borda) [inline tests/accelerators/test_dp.py:142]: isn't this already part of the model?

**Review by Borda — APPROVED** [review]

**Review by justusschock — APPROVED** [review]

File: tests/accelerators/test_dp.py (hunk 1 above)
Comment (justusschock) [reply tests/accelerators/test_dp.py:152]: yes it is, since only a copy of the model is moved in DP. Maybe we can still set the device to the root device in the plugin/wrapper?

File: tests/accelerators/test_dp.py (hunk 2 above)
Comment (carmocca) [reply tests/accelerators/test_dp.py:142]: I believe the difference is the batch size

Comment (awaelchli) [reply tests/accelerators/test_dp.py:142]: Yes exactly and because it's a DP test with 2 gpus we need at least batch size 2 for this test to be meaningful.

**Review by carmocca — APPROVED** [review]

File: tests/accelerators/test_dp.py (hunk 1 above)
Comment (awaelchli) [reply tests/accelerators/test_dp.py]: yes, already looking into that. I think it should be possible.
